### PR TITLE
Make configs and translations publishable

### PR DESCRIPTION
**Pull request by aerni**
Now you can publish an addon's configs and translations using a `$configs` and `$translations` property; The same way you can already do it with `$scripts`, `$stylesheets`, and `$publishables`.

```php
protected $configs = [
  __DIR__.'/../config/config.php',
];

protected $translations = [
  __DIR__.'/../resources/lang',
];
```

It will only publish the configs and translations that don't exist at the destination. I think this is a good practice, as you don't want to override file changes, the user may have made. If you do wish to override those files, you can still opt out of automatic publishing and use a [post-install command](https://statamic.dev/extending/addons#post-install-commands) instead.

### CODE REVIEW

**Comment by jasonvarga** [comment]
This is great but I think may be a little overkill.

I think most addons would only have one config file. We could probably just check if there's a `config/addon-slug.php`, `mergeConfigFrom` that, and `publishes` it. We should use the `addon-slug-config` tag, which is a pretty common Laravel convention. People would be able to publish the config file manually if they want, and I don't think it needs to be done by default.

Similar with translations: We could check if there's a `resources/lang` directory, `loadTranslationsFrom` that directory, and `publishes` it. Again, use the `addon-slug-translations` tag. It would handle copying ones that don't already exist in the user's app.

In both cases, they'd be free to add `--force` to the command to override everything.

We could also have `$mergeAndPublishConfig` and `$loadAndPublishTranslations` properties (defaulting to `true`). If the dev wants to do more advanced stuff, they can set them to `false` and then just wire up everything manually in their `boot` method.

**Comment by aerni** [comment]
Yeah, you're right. Might be overkill. As you said, the dev can always do it manually in the `boot` method. I'm going to tackle your proposed changes.

**Comment by aerni** [comment]
I implemented your proposed changes. The config and translations are now merged and loaded automatically. 

Publish the config: `vendor:publish --tag="addon-slug-config"`
Publish the translations: `vendor:publish --tag="addon-slug-translations"`.

You can also opt-out of this and handle things yourself:
```php
protected $mergeAndPublishConfig = false;
protected $loadAndPublishTranslations = false;
```

File: src/Providers/AddonServiceProvider.php [hunk 1]
```diff
@@ -160,6 +164,45 @@ protected function bootScripts()
...
+
+        $package = $this->getAddon()->packageName();
+        $slug = $this->getAddon()->slug();
+        $directory = $this->getAddon()->directory();
+        $origin = "{$directory}resources/lang";
+
+        $this->loadTranslationsFrom($origin, $slug);
+
+        $this->publishes([
+            $origin => resource_path("lang/vendor/{$package}")
```
Comment (jasonvarga) [inline src/Providers/AddonServiceProvider.php:200]: This should still just be slug.

**Comment by jesseleite** [comment]
I think we could automatically load translations, but publishing localizations should be opt-in, not opt-out.  I could get behind publishing the config file by default though.  If so, could simplify the boolean prop names:

```php
protected $publishTranslations = false;
protected $publishConfig = true;
```

Maybe I'm wrong, but I'd assume most people probably don't want to maintain local copies of translation files as things are changed in core.  Only publish if you intend to customize them in your app.  I'd think anyway?

**Comment by aerni** [comment]
Both the config and translations don't get published automatically. They only get merged/loaded and tagged for publishing. The user is in control if he wants to publish them by calling `vendor:publish --tag="addon-slug-{config/translations}"`

**Comment by jesseleite** [comment]
Oh, I assumed by those boolean prop names that it _was_ automatically publishing...

```php
protected $mergeAndPublishConfig = true;
protected $loadAndPublishTranslations = true;
```

**Comment by jesseleite** [comment]
👍  Oh gotcha, I just assumed by your prop names that it was automatically publishing.

**Comment by aerni** [comment]
It's derived from Laravel. They call it `$this->publishes` as well. Though it doesn't just "publish" things by itself, but needs the user to call it.

I understand the confusion. Thinking from a user perspective, I'd probably assume the same. Do you have a better idea for the prop names?

**Comment by jesseleite** [comment]
Yeah I'm aware how Laravel does it, I just skimmed your code sorry!  As for clearer property names, maybe...

```php
protected $bootTranslations = true;
protected $bootConfig = true;
```

And make `bootConfig()` method singular, to go along with Jason's suggestion. 

Thoughts?

**Comment by jasonvarga** [comment]
If the `andPublish` properties are unclear, maybe we just go with `$config` and `$translations` booleans.

You can interpret it as "do whatever automatic stuff Statamic says it'll do". I don't think the `boot` prefix is needed.

**Comment by jesseleite** [comment]
👍  Even better.

**Comment by aerni** [comment]
Yeah, that could work. I still prefer the names as they are right now. It's clearer what it does. What does @jasonvarga think?

I'll make that method singular. Oversaw this.

**Comment by aerni** [comment]
I changed the props to `config` and `translations`. Now it's coherent with props like `scripts` as well.

**Comment by aerni** [comment]
I opened a new PR #2182 because I had to change the target branch.

**Pull request closed without merging**